src/Draco.Lsp/Server/LanguageServerConnection.cs

```diff
@@ -0,0 +1,604 @@
+
+        // When the server is started, we post DeserializeFromTransport() to this block.
+        // This will start deserializing messages from the client and pushing them through the dataflow network. 
+        this.messageParser = new TransformManyBlock<IAsyncEnumerable<LspMessage>, LspMessage>(
+            ae => ae,
+            new()
+            {
+                MaxDegreeOfParallelism = DataflowBlockOptions.Unbounded
+            });
+
+        // Requests and notifications are sent to one of two blocks, the concurrent block or the exclusive block, based on whether the request processing
+        // is expected to mutate the state of the workspace. These blocks are configured using ConcurrentExclusiveSchedulerPair so that messages sent
+        // to the exclusive block are processed one at a time, in order, and never simultaneously with messages sent to the concurrent block. The concurrent block
+        // block will accept and process messages in parallel as long as no mutating messages are waiting to be processed.
+        // This allows us to use parallel processing for non-mutating ("query") requests while ensuring that mutating requests are processed in the order that the
+        // client sends them.
+        //
+        // (There are likely still race conditions that can lead to messages being processed in the wrong order, but someone would need
+        // to study the code in TPL Dataflow to work out all the edge cases. However, it's clear that the current situation is better than allowing all requests
+        // to execute in parallel, which corrupts the state of the language server very quickly.)
+        var concurrentMessageHandler = new TransformBlock<LspMessage, LspMessage?>(
+            this.ProcessRequestOrNotification,
+            new()
+            {
+                TaskScheduler = scheduler.ConcurrentScheduler,
+                MaxDegreeOfParallelism = DataflowBlockOptions.Unbounded,
+                EnsureOrdered = false,
+            });
+
+        var exclusiveMessageHandler = new TransformBlock<LspMessage, LspMessage?>(
+            this.ProcessRequestOrNotification,
+            new()
+            {
+                TaskScheduler = scheduler.ExclusiveScheduler
+                // By default, blocks execute messages in order and with no parallelism.
+            });
+
+        // ProcessRequestOrNotification returns a non-null result when a response message needs to be sent back to the client.
+        // We keep only non-null responses and forward them to SerializeToTransport, which will serialize each LspMessage object it consumes to the output stream.
+        var filterNullResponses = new TransformManyBlock<LspMessage?, LspMessage>(
+            m => m.HasValue ? new[] { m.Value } : Enumerable.Empty<LspMessage>());
+
+        this.outgoingMessages = new ActionBlock<LspMessage>(this.SerializeToTransport!);
+
+        bool ShouldHandle(LspMessage message, bool mutating)
```

Comment:
I see what you are doing here, but for me it took a sec to click. I have no idea how to write it nicer

Reply:
I've changed the name of `mutating`, hopefully it's clearer. Yeah, the conditions are not super obvious...